ddl/db_test.go

```diff
@@ -1846,6 +1846,109 @@ func (s *testDBSuite2) TestTableForeignKey(c *C) {
+	s.tk.MustExec("create table t3 (a int, b int generated always as (a+1) stored);")
+	s.tk.MustExec("alter table t3 add foreign key (b) references t2(a);")
+	s.tk.MustExec("drop table t1, t2, t3;")
+
+	// foreign key constraint can reference a stored generated column.
+	s.tk.MustExec("create table t1 (a int, b int generated always as (a+1) stored primary key);")
+	s.tk.MustExec("create table t2 (a int, foreign key (a) references t1(b));")
+	s.tk.MustExec("create table t3 (a int);")
+	s.tk.MustExec("alter table t3 add foreign key (a) references t1(b);")
+	s.tk.MustExec("drop table t1, t2, t3;")
+
+	// rejected FK options on stored generated columns
+	s.tk.MustGetErrCode("create table t1 (a int, b int generated always as (a+1) stored, foreign key (b) references t2(a) on update set null);", tmysql.ErrWrongFKOptionForGeneratedColumn)
+	s.tk.MustGetErrCode("create table t1 (a int, b int generated always as (a+1) stored, foreign key (b) references t2(a) on update cascade);", tmysql.ErrWrongFKOptionForGeneratedColumn)
+	s.tk.MustGetErrCode("create table t1 (a int, b int generated always as (a+1) stored, foreign key (b) references t2(a) on update set default);", tmysql.ErrWrongFKOptionForGeneratedColumn)
+	s.tk.MustGetErrCode("create table t1 (a int, b int generated always as (a+1) stored, foreign key (b) references t2(a) on delete set null);", tmysql.ErrWrongFKOptionForGeneratedColumn)
+	s.tk.MustGetErrCode("create table t1 (a int, b int generated always as (a+1) stored, foreign key (b) references t2(a) on delete set default);", tmysql.ErrWrongFKOptionForGeneratedColumn)
+	s.tk.MustExec("create table t2 (a int primary key);")
+	s.tk.MustExec("create table t1 (a int, b int generated always as (a+1) stored);")
+	s.tk.MustGetErrCode("alter table t1 add foreign key (b) references t2(a) on update set null;", tmysql.ErrWrongFKOptionForGeneratedColumn)
+	s.tk.MustGetErrCode("alter table t1 add foreign key (b) references t2(a) on update cascade;", tmysql.ErrWrongFKOptionForGeneratedColumn)
+	s.tk.MustGetErrCode("alter table t1 add foreign key (b) references t2(a) on update set default;", tmysql.ErrWrongFKOptionForGeneratedColumn)
+	s.tk.MustGetErrCode("alter table t1 add foreign key (b) references t2(a) on delete set null;", tmysql.ErrWrongFKOptionForGeneratedColumn)
+	s.tk.MustGetErrCode("alter table t1 add foreign key (b) references t2(a) on delete set default;", tmysql.ErrWrongFKOptionForGeneratedColumn)
+	s.tk.MustExec("drop table t1, t2;")
+
+	// special case: TiDB error different from MySQL 8.0
+	// MySQL: ERROR 3104 (HY000): Cannot define foreign key with ON UPDATE SET NULL clause on a generated column.
+	// TiDB:  ERROR 1146 (42S02): Table 'test.t2' doesn't exist
+	s.tk.MustExec("create table t1 (a int, b int generated always as (a+1) stored);")
+	s.tk.MustGetErrCode("alter table t1 add foreign key (b) references t2(a) on update set null;", tmysql.ErrNoSuchTable)
```

Comment:
The MySQL error seems weird, how does MySQL know the `t2(a)` is a generated column if the table `t2` does not exist.

Reply:
@lonng Here it means `t1.b` is a generated column, rather `t2.a`.